lib/IDE/Refactoring.cpp

```diff
@@ -2752,78 +2752,107 @@ bool RefactoringActionLocalizeString::performChange() {
   return false;
 }
 
+struct MemberwiseParameter {
+  Identifier Name;
+  Type MemberType;
+  Expr *DefaultExpr;
+
+  MemberwiseParameter(Identifier name, Type type, Expr *initialExpr)
+    : Name(name), MemberType(type), DefaultExpr(initialExpr) {}
+};
+
 static void generateMemberwiseInit(SourceEditConsumer &EditConsumer,
-                            SourceManager &SM,
-                            SmallVectorImpl<std::string>& memberNameVector,
-                            SmallVectorImpl<std::string>& memberTypeVector,
-                            SourceLoc targetLocation) {
-  
-  assert(!memberTypeVector.empty());
-  assert(memberTypeVector.size() == memberNameVector.size());
+                                   SourceManager &SM,
+                                   ArrayRef<MemberwiseParameter> memberVector,
+                                   SourceLoc targetLocation) {
   
+  assert(!memberVector.empty());
+
   EditConsumer.accept(SM, targetLocation, "\ninternal init(");
-  
-  for (size_t i = 0, n = memberTypeVector.size(); i < n ; i++) {
-    EditConsumer.accept(SM, targetLocation, memberNameVector[i] + ": " +
-                        memberTypeVector[i]);
-    
-    if (i != memberTypeVector.size() - 1) {
-      EditConsumer.accept(SM, targetLocation, ", ");
+  auto insertMember = [&SM](const MemberwiseParameter &memberData,
+                            llvm::raw_ostream &OS, bool wantsSeparator) {
+    OS << memberData.Name << ": " << memberData.MemberType.getString();
+    if (auto *expr = memberData.DefaultExpr) {
+      if (isa<NilLiteralExpr>(expr)) {
```

Comment:
This whole check feels like it should be factored out and maybe made into something that `ParamDecl::getDefaultValueStringRepresentation` could use, in general.

As it stands, this is duplicating the whole purpose behind `DefaultArgumentKind`.

Reply:
This can be a follow-up patch. I was going to tackle this very soon too.

Reply:
Feels weird to allocate and throw away an entire ParamDecl here, but I see what you mean about sequestering the logic up in the AST - I think that primitive on ParamDecl can be generalized to accept this data instead.  Or the header with the convenience for printing default arguments can actually be moved to the includes directory so I can consume it here.  Didn’t want to touch any of that for such a simple patch.

Reply:
Yeah, it should be factored out of ParamDecl. I don’t think it should block this patch.